apps/aecore/lib/aecore/peers/sync.ex

```diff
@@ -0,0 +1,606 @@
+  def new_peer?(peer_pid, header, agreed_height, hash) do
+    GenServer.call(__MODULE__, {:is_new_peer, self(), peer_pid, header, agreed_height, hash})
+  end
+
+  @doc """
+  Does the fetching of the blocks depending on the hash_pool.
+  """
+  @spec fetch_next(pid(), non_neg_integer(), binary(), any()) :: :done | tuple()
+  def fetch_next(peer_pid, height_in, hash_in, result) do
+    GenServer.call(__MODULE__, {:fetch_next, peer_pid, height_in, hash_in, result}, 30_000)
+  end
+
+  @spec update_hash_pool(list()) :: list(Header.t())
+  def update_hash_pool(hashes) do
+    GenServer.call(__MODULE__, {:update_hash_pool, hashes})
+  end
+
+  def delete_from_pool(peer_pid) do
+    GenServer.cast(__MODULE__, {:delete_from_pool, peer_pid})
+  end
+
+  ## INNER FUNCTIONS ##
+
+  def handle_call(:state, _from, state) do
+    {:reply, state, state}
+  end
+
+  def handle_call({:start_sync, peer_pid, remote_hash}, _from, state) do
+    :ok = Jobs.enqueue(:sync_jobs, {:start_sync, peer_pid, remote_hash})
+    {:reply, :ok, state}
+  end
+
+  def handle_call({:fetch_mempool, peer_pid}, _from, state) do
+    :ok = Jobs.enqueue(:sync_jobs, {:fetch_mempool, peer_pid})
+    {:reply, :ok, state}
+  end
+
+  def handle_call(
+        {:is_new_peer, process_pid, peer_pid, header, agreed_height, hash},
+        _from,
+        %{sync_pool: pool} = state
+      ) do
+    height = header.height
+    difficulty = Scientific.target_to_difficulty(header.target)
+
+    {new?, new_pool} =
+      try_add_peer(
+        Sync.new(%{
+          difficulty: difficulty,
+          from: agreed_height,
+          to: height,
+          hash: hash,
+          peer: peer_pid,
+          pid: process_pid
+        }),
+        pool
+      )
+
+    if new? do
+      Process.monitor(process_pid)
+    else
+      :ok
+    end
+
+    {:reply, new?, %{state | sync_pool: new_pool}}
+  end
+
+  def handle_call({:sync_in_progress, peer_pid}, _from, %{sync_pool: pool} = state) do
+    result =
+      case Enum.find(pool, false, fn peer -> Map.get(peer, :pid) == peer_pid end) do
+        false ->
+          false
+
+        peer ->
+          {true, peer}
+      end
+
+    {:reply, result, state}
+  end
+
+  def handle_call({:update_hash_pool, hashes}, _from, state) do
+    hash_pool = merge(state.hash_pool, hashes)
+    Logger.debug(fn -> "Hash pool now contains #{inspect(hash_pool)} hashes" end)
+    {:reply, :ok, %{state | hash_pool: hash_pool}}
+  end
+
+  def handle_call({:fetch_next, peer_pid, height_in, hash_in, result}, _from, state) do
+    hash_pool =
+      case result do
+        {:ok, block} ->
+          block_height = block.header.height
+          block_hash = BlockValidation.block_header_hash(block.header)
+
+          # If the hash of this block does not fit wanted hash, it is discarded
+          # (In case we ask for block with hash X and we get a block with hash Y)
+          List.keyreplace(
```

Comment:
Also: Is there some reason for state.hash_pool to be a list instead of map ({block_height, block_hash} -> block | peer)?

Reply:
Same as my previous comment. `epoch` did it like so, and I don't see a reason why to change it. If, for some reason, we decide to do so, will be issues a separate issue for this change.

Reply:
The goal is to copy specification, not translate erlang code to elixir. The question from @cytadela8 seems appropriate.